state/snapshotsManager.go

```diff
@@ -0,0 +1,454 @@
+		WaitTimeForSnapshotEpochCheck: waitTimeForSnapshotEpochCheck,
+		SnapshotWaitTimeout:           snapshotWaitTimeout,
+	})
+	if err != nil {
+		log.Error("error waiting for storage epoch change", "err", err)
+		sm.earlySnapshotCompletion(stats)
+		return
+	}
+
+	if !trieStorageManager.ShouldTakeSnapshot() {
+		log.Debug("skipping snapshot", "rootHash", rootHash, "epoch", epoch)
+
+		sm.earlySnapshotCompletion(stats)
+		return
+	}
+
+	missingNodesChannel := make(chan []byte, missingNodesChannelSize)
+	iteratorChannels := sm.channelsProvider.GetIteratorChannels()
+
+	sm.stateMetrics.UpdateMetricsOnSnapshotStart()
+
+	go func() {
+		stats.NewSnapshotStarted()
+
+		trieStorageManager.TakeSnapshot("", rootHash, rootHash, iteratorChannels, missingNodesChannel, stats, epoch)
+		sm.snapshotUserAccountDataTrie(true, rootHash, iteratorChannels, missingNodesChannel, stats, epoch, trieStorageManager)
+
+		stats.SnapshotFinished()
+	}()
+
+	go sm.syncMissingNodes(missingNodesChannel, iteratorChannels.ErrChan, stats, sm.getTrieSyncer())
+
+	go sm.processSnapshotCompletion(stats, trieStorageManager, missingNodesChannel, iteratorChannels.ErrChan, rootHash, epoch)
+}
+
+func (sm *snapshotsManager) earlySnapshotCompletion(stats *snapshotStatistics) {
+	sm.mutex.Lock()
+	defer sm.mutex.Unlock()
+
+	stats.SnapshotFinished()
+	sm.isSnapshotInProgress.Reset()
+}
+
+func (sm *snapshotsManager) waitForStorageEpochChange(args StorageEpochChangeWaitArgs) error {
+	if args.SnapshotWaitTimeout < args.WaitTimeForSnapshotEpochCheck {
+		return fmt.Errorf("timeout (%s) must be greater than wait time between snapshot epoch check (%s)", args.SnapshotWaitTimeout, args.WaitTimeForSnapshotEpochCheck)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), args.SnapshotWaitTimeout)
+	defer cancel()
+
+	timer := time.NewTimer(args.WaitTimeForSnapshotEpochCheck)
+	defer timer.Stop()
+
+	for {
+		timer.Reset(args.WaitTimeForSnapshotEpochCheck)
+
+		if args.TrieStorageManager.IsClosed() {
+			return core.ErrContextClosing
+		}
+
+		latestStorageEpoch, err := args.TrieStorageManager.GetLatestStorageEpoch()
+		if err != nil {
+			return err
+		}
+
+		if latestStorageEpoch == args.Epoch {
+			return nil
+		}
+
+		select {
+		case <-timer.C:
+			continue
+		case <-ctx.Done():
+			return fmt.Errorf("timeout waiting for storage epoch change, snapshot epoch %d", args.Epoch)
+		}
+	}
+}
+
+func (sm *snapshotsManager) snapshotUserAccountDataTrie(
+	isSnapshot bool,
+	mainTrieRootHash []byte,
+	iteratorChannels *common.TrieIteratorChannels,
+	missingNodesChannel chan []byte,
+	stats common.SnapshotStatisticsHandler,
+	epoch uint32,
+	trieStorageManager common.StorageManager,
+) {
+	if iteratorChannels.LeavesChan == nil {
+		return
+	}
+
+	for leaf := range iteratorChannels.LeavesChan {
+		userAccount, skipAccount, err := getUserAccountFromBytes(sm.accountFactory, sm.marshaller, leaf.Key(), leaf.Value())
+		if skipAccount || err != nil {
+			continue
```

Comment:
you've changed the behavior here. Is this correct? In the previous version, we would have called this: `iteratorChannels.ErrChan.WriteInChanNonBlocking(err)`

Reply:
Missed that. Fixed.